src/main/kotlin/org/wycliffeassociates/resourcecontainer/ZipAccessor.kt

```diff
@@ -0,0 +1,92 @@
+package org.wycliffeassociates.resourcecontainer
+
+import java.io.File
+import java.io.IOException
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.io.FileOutputStream
+import java.util.zip.ZipEntry
+import java.util.zip.ZipFile
+import java.util.zip.ZipOutputStream
+
+class ZipAccessor(private val file: File) : IResourceContainerAccessor {
+    private var _zipFile: ZipFile? = null
+
+    private fun openZipFile(): ZipFile {
+        if (_zipFile == null) _zipFile = ZipFile(file)
+        return _zipFile as ZipFile
+    }
+
+    private fun closeZipFile() {
+        _zipFile?.let(ZipFile::close)
+        _zipFile = null
+    }
+
+    @Synchronized
+    override fun getReader(filename: String): BufferedReader {
+        return openZipFile()
+                .getInputStream(openZipFile().getEntry(filename))
+                .bufferedReader()
+    }
+
+    @Synchronized
+    override fun checkFileExists(filename: String): Boolean {
+        return openZipFile()
+                .entries()
+                .asSequence()
+                .map { it.name }
+                .contains(filename)
+    }
+
+    override fun initWrite() {
+        // noop
+    }
+
+    @Synchronized
+    override fun write(filename: String, writeFcn: (BufferedWriter) -> Unit) {
+        val doCopy = file.exists()
+        val dest = when (doCopy) {
+            true -> File.createTempFile("otter", ".zip", file.parentFile)
+            false -> file
+        }
+        ZipOutputStream(FileOutputStream(dest)).use { zos ->
+            var found = false
+            if (doCopy) {
+                openZipFile().entries().iterator().forEach {
+                    if (it.name == filename) {
+                        zos.write(ZipEntry(filename), writeFcn)
+                        found = true
+                    } else {
+                        // Simply doing zos.putNextEntry(it) resulted in ZipExceptions - invalid entry
+                        // compressed size
+                        val destEntry = ZipEntry(it.name)
+                        zos.putNextEntry(destEntry)
+                        openZipFile().getInputStream(destEntry).use { inStream -> inStream.copyTo(zos) }
+                        zos.tryCloseEntry()
+                    }
+                }
+            }
+            if (!found) {
+                zos.write(ZipEntry(filename), writeFcn)
+            }
+        }
+        if (doCopy) {
+            closeZipFile()
+            file.delete()
+            dest.renameTo(file)
+        }
+    }
+}
+
+private fun ZipOutputStream.write(zipEntry: ZipEntry, writeFcn: (BufferedWriter) -> Unit) {
+    putNextEntry(zipEntry)
+    writeFcn(bufferedWriter())
+    tryCloseEntry()
```

Comment:
this wouldn't execute if an exception was thrown previously

Reply:
Good point.  This is better handed using `try/finally`, with the close call in the finally.

Reply:
done